File: c++/leetcode/array/3Sum.cpp
#include <algorithm>
#include <vector>

using namespace std;
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        vector<vector<int>> result;

        if (nums.size() < 3) {
            return result;
        }

        sort(nums.begin(),nums.end());

        for (int i=0; i<nums.size(); ++i) {
            if (i>0 && nums[i]==nums[i-1]) {
                continue;
            }

            int third = nums.size()-1;
            int target = -nums[i];

            for (int j=i+1; j<nums.size(); j++) {
                if (j>i+1 && nums[j] == nums[j-1]) {
                    continue;
                }

                while(j<third && nums[j]+nums[third] > target){
                    third--;
                }

                if (third == j) {
                    break;
                }

                if(nums[j]+nums[third] == target){
                    result.push_back({nums[i],nums[j],nums[third]});
                }
            }
        }
        return result;
    }
};

File: c++/leetcode/array/3Sum.go
package main

import (
	"sort"
)

func threeSum(nums []int) [][]int {

	if len(nums) < 3 {
		return [][]int{}
	}

	result := [][]int{}
	sort.Ints(nums)

	for i := 0; i <= len(nums)-3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		k := len(nums) - 1
		target := -1 * nums[i]

		for j := i + 1; j <= len(nums)-2; j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}

			for j < k && nums[j]+nums[k] > target {
				k--
			}

			if j == k {
				break
			}

			if nums[j]+nums[k] == target {
				result = append(result, []int{nums[i], nums[j], nums[k]})
			}

		}
	}
	return result
}
